src/manager/evm/manager.rs

```diff
@@ -834,6 +856,124 @@ impl EthSubnetManager {
     }
 }
 
+/// Receives an input `FunctionCall` and returns a new instance
+/// after estimating an optimal `gas_premium` for the transaction
+async fn call_with_premium_estimation<B, D, M>(
+    signer: Arc<DefaultSignerMiddleware>,
+    call: ethers_contract::FunctionCall<B, D, M>,
+) -> Result<ethers_contract::FunctionCall<B, D, M>>
+where
+    B: std::borrow::Borrow<D>,
+    M: ethers::abi::Detokenize,
+{
+    let (max_priority_fee_per_gas, _) = premium_estimation(signer).await?;
+    Ok(call.gas_price(max_priority_fee_per_gas))
+}
+
+/// Returns an estimation of an optimal `gas_premium` and `gas_fee_cap`
+/// for a transaction considering the average premium, base_fee and reward percentile from
+/// past blocks
+/// This is adaptation of ethers' `eip1559_default_estimator`:
+/// https://github.com/gakonst/ethers-rs/blob/5dcd3b7e754174448f9a8cbfc0523896609629f9/ethers-core/src/utils/mod.rs#L476
+async fn premium_estimation(
+    signer: Arc<DefaultSignerMiddleware>,
+) -> Result<(ethers::types::U256, ethers::types::U256)> {
+    let base_fee_per_gas = signer
+        .get_block(ethers::types::BlockNumber::Latest)
+        .await?
+        .ok_or_else(|| anyhow!("Latest block not found"))?
+        .base_fee_per_gas
+        .ok_or_else(|| anyhow!("EIP-1559 not activated"))?;
+
+    let fee_history = signer
+        .fee_history(
+            ethers::utils::EIP1559_FEE_ESTIMATION_PAST_BLOCKS,
+            ethers::types::BlockNumber::Latest,
+            &[ethers::utils::EIP1559_FEE_ESTIMATION_REWARD_PERCENTILE],
+        )
+        .await?;
+
+    let max_priority_fee_per_gas = estimate_priority_fee(fee_history.reward); //overestimate?
+    let potential_max_fee = base_fee_surged(base_fee_per_gas);
+    let max_fee_per_gas = if max_priority_fee_per_gas > potential_max_fee {
+        max_priority_fee_per_gas + potential_max_fee
+    } else {
+        potential_max_fee
+    };
+
+    Ok((max_priority_fee_per_gas, max_fee_per_gas))
+}
+
+/// Implementation borrowed from
+/// https://github.com/gakonst/ethers-rs/blob/ethers-v2.0.8/ethers-core/src/utils/mod.rs#L582
+/// Refer to the implementation for unit tests
+fn base_fee_surged(base_fee_per_gas: U256) -> U256 {
+    if base_fee_per_gas <= U256::from(40_000_000_000u64) {
+        base_fee_per_gas * 2
+    } else if base_fee_per_gas <= U256::from(100_000_000_000u64) {
+        base_fee_per_gas * 16 / 10
+    } else if base_fee_per_gas <= U256::from(200_000_000_000u64) {
+        base_fee_per_gas * 14 / 10
+    } else {
+        base_fee_per_gas * 12 / 10
+    }
+}
```

Comment:
This is some quality logic 😓 (I realise it's borrowed)

Any concerns re how these constants translate to Filecoin vs eth? It's probably fine.